fe/fe-core/src/main/java/com/starrocks/transaction/DatabaseTransactionMgr.java

```diff
@@ -739,40 +739,51 @@ public List<TransactionStateBatch> getReadyToPublishTxnListBatch() {
                         Config.lake_batch_publish_min_version_num,
                         Config.lake_batch_publish_max_version_num, txnId);
                 List<TransactionState> states = txnsWithDependency.stream().map(id -> idToRunningTransactionState.get(id))
+                        .filter(Objects::nonNull)
                         .collect(Collectors.toList());
+                if (states.isEmpty()) {
+                    continue;
+                }
+                if (states.size() == 1) { // fast path: no batch
+                    result.add(new TransactionStateBatch(states));
+                    continue;
+                }
+
+                // Only single table transactions will be batched together.
+                Preconditions.checkState(states.get(0).getTableIdList().size() == 1);
+
+                long tableId = states.get(0).getTableIdList().get(0);
+
                 // check whether version is consequent
                 // for schema change will occupy a version
                 Map<Long, PartitionCommitInfo> versions = new HashMap<>();
-                long tableId = -1;
-                if (states.size() != 0) {
-                    tableId = states.get(0).getTableIdList().get(0);
-                    versions.putAll(states.get(0).getTableCommitInfo(tableId).getIdToPartitionCommitInfo());
-                }
 
-                boolean consecutive = true;
-                for (int i = 1; i < states.size() && consecutive; i++) {
+                outerLoop:
+                for (int i = 0; i < states.size(); i++) {
                     TransactionState state = states.get(i);
-                    for (Map.Entry<Long, PartitionCommitInfo> item :
-                            state.getTableCommitInfo(tableId).getIdToPartitionCommitInfo().entrySet()) {
-                        if (versions.containsKey(item.getKey())) {
+                    TableCommitInfo tableInfo = state.getTableCommitInfo(tableId);
+                    // TableCommitInfo could be null if the table has been dropped before this transaction is committed.
+                    if (tableInfo == null) {
```

Comment:
Since the table has been dropped, and all transactionStates involve the same table, can we finish all these thansactions short-circuit instead of converting to single publish to deal with the situation？

Reply:
Doing so requires some modifications of the batch pulishing code, because some places called `TransactionState.getTableCommitInfo()` without null checking.